packages/cozy-konnector-libs/src/libs/saveBills.js

```diff
@@ -197,4 +203,68 @@ function checkRequiredAttributes(entries) {
     }
   }
 }
+
+function appendContractsDatas(tempEntries, options) {
+  if (options.contractLabel && options.contractId === undefined) {
+    log('warn', 'contractLabel used without contractId')
+    return tempEntries
+  }
+  let newEntries = tempEntries
+  if (options.contractId) {
+    // if contractId passed by option
+    newEntries = appendContractDatasByOption(newEntries, options)
+  } else if (billsHaveContractId(newEntries)) {
+    // if contractId passed by bill attribute
+    newEntries = appendContractDatasByAttribute(newEntries, options)
+  }
+  return newEntries
+}
+
+function appendContractDatasByOption(newEntries, options) {
+  // Surcharge label if not define
+  if (!options.contractLabel) {
+    options.contractLabel = options.contractId
+  }
+  options.subPath = options.contractLabel
+
+  // Add contractId to deduplication keys
+  if (options.keys) {
+    options.keys.push('contractId')
+  }
+
+  // Add datas in bills
+  for (let entry of newEntries) {
+    entry.contractLabel = options.contractLabel
+    entry.contractId = options.contractId
+  }
+  return newEntries
```

Comment:
This function does 

- defaulting of values inside options
- iteration
- assigning values from options inside an entry

Those 3 things are decoupled and should not be done inside the same function.

This function looks very similar to `appendContractDatasByAttribute`, it is not clear at first sight what is the difference.

IMHO there should be 

- one function taking contractData and merging it with an entry => mergeContractDataIntoEntry
- one function making the contractData from the options
- one function doing the iteration entries = entries.map(entry => mergeContractDataIntoEntry(contractData, entry))

It seems you have to modify options.keys at some point to add contractId to the deduplication keys. It seems dangerous to do this implicitly by changing options, the caller
of appendContractsDatas might not now that options has changed.